File: src/main/kotlin/xyz/mahmoudahmed/dsl/NestedSimpleMapping.java
package xyz.mahmoudahmed.dsl;

import xyz.mahmoudahmed.adapter.DataNode;

import java.util.function.Function;
import java.util.function.Predicate;

public class NestedSimpleMapping implements MappingRule {
    private final String sourcePath;
    private final String targetPath;
    private final Function<Object, Object> transformation;
    private final Predicate<Object> condition;

    NestedSimpleMapping(String sourcePath, String targetPath,
                        Function<Object, Object> transformation, Predicate<Object> condition) {
        this.sourcePath = sourcePath;
        this.targetPath = targetPath;
        this.transformation = transformation;
        this.condition = condition;
    }

    @Override
    public void apply(MappingContext context, Object target) {
        if (condition != null && !condition.test(context.getSourceData())) {
            return;
        }

        Object sourceValue = context.getValueByPath(sourcePath);
        if (sourceValue == null) {
            return;
        }

        Object targetValue = sourceValue;
        if (transformation != null) {
            targetValue = transformation.apply(sourceValue);
        }

        if (target instanceof DataNode.ObjectNode) {
            SimpleMapping.setValueInDataNode((DataNode.ObjectNode) target, targetPath, targetValue);
        }
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/collections/NestedSimpleMapping.kt
package xyz.mahmoudahmed.dsl.collections

import xyz.mahmoudahmed.adapter.DataNode
import xyz.mahmoudahmed.dsl.core.MappingContext
import xyz.mahmoudahmed.dsl.core.MappingRule
import xyz.mahmoudahmed.dsl.core.SimpleMapping

/**
 * Mapping rule for nested mappings within collections.
 */
class NestedSimpleMapping<T: Any>(
    private val sourcePath: String,
    private val targetPath: String,
    private val transformation: ((T) -> Any)?,
    private val condition: ((Any) -> Boolean)?
) : MappingRule {

    @Suppress("UNCHECKED_CAST")
    override fun apply(context: MappingContext, target: Any) {
        if (condition != null && !condition.invoke(context.sourceData)) {
            return
        }

        val sourceValue = context.getValueByPath(sourcePath) ?: return

        val targetValue = transformation?.invoke(sourceValue as T) ?: sourceValue

        if (target is DataNode.ObjectNode) {
            SimpleMapping.setValueInDataNode(target, targetPath, targetValue)
        }
    }
}